<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="generator" content="pandoc">
    <title>EBI Intro RNA-seq: Introduction to Next-generation Sequencing</title>
    <link rel="shortcut icon" type="image/x-icon" href="/favicon.ico" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="css/swc.css" />
    <link rel="alternate" type="application/rss+xml" title="EBI Intro RNA-Seq 2022" href="https://www.ebi.ac.uk/training/events/introduction-rna-seq-and-functional-interpretation-2022/"/>
    <meta charset="UTF-8" />
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body class="lesson">
    <div class="container card">
      <div class="banner">
        <a href="https://liz-fernandez.github.io//" title="EBI Intro RNA-Seq 2022">
          <img alt="Logo EBI-Cinvestav" src="img/EBI-cinvestav.jpg" />
        </a>
      </div>
      <article>
      <div class="row">
        <div class="col-md-10 col-md-offset-1">
                    <a href="index.html"><h1 class="title">Introduction to Next-generation Sequencing</h1></a>
          <h2 class="subtitle">De novo transcriptome assembly</h2>
          <section class="objectives panel panel-warning">
<div class="panel-heading">
<h2 id="learning-objectives"><span class="glyphicon glyphicon-certificate"></span>Learning Objectives</h2>
</div>
<div class="panel-body">
<ul>
<li>Learn how the <em>de novo</em> RNA-Seq assembly works.</li>
<li>Learn how to use Trinity to assemble a transcriptome <em>de novo</em>.</li>
</ul>
</div>
</section>
<p>We will use the reads from four experiments in ‘/usr/local/data/’</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">ls</span> Sp*fq.gz</code></pre></div>
<p>Let’s explore the reads in each file using a ‘for’ cycle:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">for</span> <span class="kw">fastq</span> in S*fq.gz <span class="kw">;</span> <span class="kw">do</span> <span class="kw">echo</span> <span class="ot">$fastq</span><span class="kw">;</span> <span class="kw">zcat</span> <span class="ot">$fastq</span> <span class="kw">|</span> <span class="kw">head</span> <span class="kw">;</span> <span class="kw">wait</span> <span class="kw">;</span> <span class="kw">done</span> </code></pre></div>
<p>This command will show us the first 10 lines of each file in an iterative way. This structure is known as a <code>for loop</code>. A for loop allows us to execute a command in several files sequentially. It is extremely useful when we have several files.</p>
<p>Once we verify that the data is correct, We will use the Trinity program to assemble the transcripts. In this exercise we are assuming that these files were already made filtered by quality.</p>
<p>And start using a generic command:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">Trinity</span> --seqType fq --SS_lib_type RF \
--left Sp_log.left.fq.gz,Sp_hs.left.fq.gz \
--right Sp_log.right.fq.gz,Sp_hs.right.fq.gz \
--CPU 2 --max_memory 1G --no_normalize_reads</code></pre></div>
<p>This command will take approximately 15 minutes to assemble a transcriptome.</p>
<p>The options (flags) that we have used are the following:</p>
<ul>
<li>–seqType fq - We indicate that we are using fastq type files</li>
<li>–SS_lib_type RF - We indicate that the library was built using paired-end reads in the RF orientation (reverse-forward)</li>
<li>–left - Left side reads (o R1)</li>
<li>–right - Right side reads (o R2)</li>
<li>–CPU 2 - Use 2 CPUs</li>
<li>–max_memory 1G - Indicate that the maximum RAM to use is 1GB</li>
<li>–no_normalize_reads - Do not carry out digital normalization of reads</li>
</ul>
<p>These are the most essential options to carry out the analysis. Therefore it is very important to know the orientation of the library which depends on the experimental protocol that was used to generate it.</p>
<aside class="callout panel panel-info">
<div class="panel-heading">
<h2 id="library-strand-orientation"><span class="glyphicon glyphicon-pushpin"></span>Library strand orientation</h2>
</div>
<div class="panel-body">
<p>The library orientation is an essential parameter for the assembly, since this will dictate the orientation of the transcripts.</p>
<p>This is a key that will allow you to decide which parameter to use to indicate to Trinity the orientation of its library.</p>
<div class="figure">
<img src="fig/strand_specificity.jpg" alt="Library orientation" />
<p class="caption">Library orientation</p>
</div>
</div>
</aside>
<section class="challenge panel panel-success">
<div class="panel-heading">
<h2 id="why-do-we-mix-libraries"><span class="glyphicon glyphicon-pencil"></span>Why do we mix libraries?</h2>
</div>
<div class="panel-body">
<p>In the command used at the top we can see that We are assembling the transcriptome with the log library as well as the library hs. Why not assemble them independently? What advantages or disadvantages do you think it would have assemble them together or separately?</p>
</div>
</section>
<p>One of the common problems with Trinity is the lack of memory. A typical Trinity analysis (without digital normalization) requires ~1 hour and ~1GB of RAM for ~1 million paired-end reads. This is why these analyzes are carried out on a server with large amounts of memory where it can be left running for several days</p>
<p>Trinity has many other options which we can explore writing (in a different terminal than the one we are using for our analysis):</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">Trinity</span></code></pre></div>
<pre class="output"><code>################################################################################

     ______  ____   ____  ____   ____  ______  __ __
    |      ||    \ |    ||    \ |    ||      ||  |  |
    |      ||  D  ) |  | |  _  | |  | |      ||  |  |
    |_|  |_||    /  |  | |  |  | |  | |_|  |_||  ~  |
      |  |  |    \  |  | |  |  | |  |   |  |  |___, |
      |  |  |  .  \ |  | |  |  | |  |   |  |  |     |
      |__|  |__|\_||____||__|__||____|  |__|  |____/

#
#
# Required:
#
#  --seqType &lt;string&gt;      :type of reads: (&#39;fa&#39; or &#39;fq&#39;)
#
#  --max_memory &lt;string&gt;      :suggested max memory to use by Trinity where limiting can be enabled. (jellyfish, sorting, etc)
#                            provided in Gb of RAM, ie.  &#39;--max_memory 10G&#39;
#
#  If paired reads:
#      --left  &lt;string&gt;    :left reads, one or more file names (separated by commas, no spaces)
#      --right &lt;string&gt;    :right reads, one or more file names (separated by commas, no spaces)
#
#  Or, if unpaired reads:
#      --single &lt;string&gt;   :single reads, one or more file names, comma-delimited (note, if single file contains pairs, can use flag: --run_as_paired )
#
#  Or,
#      --samples_file &lt;string&gt;         tab-delimited text file indicating biological replicate relationships.
#                                   ex.
#                                        cond_A    cond_A_rep1    A_rep1_left.fq    A_rep1_right.fq
#                                        cond_A    cond_A_rep2    A_rep2_left.fq    A_rep2_right.fq
#                                        cond_B    cond_B_rep1    B_rep1_left.fq    B_rep1_right.fq
#                                        cond_B    cond_B_rep2    B_rep2_left.fq    B_rep2_right.fq
#
#                      # if single-end instead of paired-end, then leave the 4th column above empty.
#
####################################
##  Misc:  #########################
#
#  --SS_lib_type &lt;string&gt;          :Strand-specific RNA-Seq read orientation.
#                                   if paired: RF or FR,
#                                   if single: F or R.   (dUTP method = RF)
#                                   See web documentation.
#
#  --CPU &lt;int&gt;                     :number of CPUs to use, default: 2
#  --min_contig_length &lt;int&gt;       :minimum assembled contig length to report
#                                   (def=200)
#
#  --long_reads &lt;string&gt;           :fasta file containing error-corrected or circular consensus (CCS) pac bio reads
#                                   (** note: experimental parameter **, this functionality continues to be under development)
#
#  --genome_guided_bam &lt;string&gt;    :genome guided mode, provide path to coordinate-sorted bam file.
#                                   (see genome-guided param section under --show_full_usage_info)
#
#  --jaccard_clip                  :option, set if you have paired reads and
#                                   you expect high gene density with UTR
#                                   overlap (use FASTQ input file format
#                                   for reads).
#                                   (note: jaccard_clip is an expensive
#                                   operation, so avoid using it unless
#                                   necessary due to finding excessive fusion
#                                   transcripts w/o it.)
#
#  --trimmomatic                   :run Trimmomatic to quality trim reads
#                                        see &#39;--quality_trimming_params&#39; under full usage info for tailored settings.
#
#
#  --no_normalize_reads            :Do *not* run in silico normalization of reads. Defaults to max. read coverage of 50.
#                                       see &#39;--normalize_max_read_cov&#39; under full usage info for tailored settings.
#                                       (note, as of Sept 21, 2016, normalization is on by default)
#
#  --no_distributed_trinity_exec   :do not run Trinity phase 2 (assembly of partitioned reads), and stop after generating command list.
#
#
#  --output &lt;string&gt;               :name of directory for output (will be
#                                   created if it doesn&#39;t already exist)
#                                   default( your current working directory: &quot;/home/sfernandez/TEST/TRANSCRIPTOMICS/trinity_out_dir&quot;
#                                    note: must include &#39;trinity&#39; in the name as a safety precaution! )
#
#  --workdir &lt;string&gt;              :where Trinity phase-2 assembly computation takes place (defaults to --output setting).
#                                  (can set this to a node-local drive or RAM disk)
#
#  --full_cleanup                  :only retain the Trinity fasta file, rename as ${output_dir}.Trinity.fasta
#
#  --cite                          :show the Trinity literature citation
#
#  --verbose                       :provide additional job status info during the run.
#
#  --version                       :reports Trinity version (Trinity-v2.5.0) and exits.
#
#  --show_full_usage_info          :show the many many more options available for running Trinity (expert usage).
#
#
###############################################################################
#
#  *Note, a typical Trinity command might be:
#
#        Trinity --seqType fq --max_memory 50G --left reads_1.fq  --right reads_2.fq --CPU 6
#
#
#    and for Genome-guided Trinity:
#
#        Trinity --genome_guided_bam rnaseq_alignments.csorted.bam --max_memory 50G
#                --genome_guided_max_intron 10000 --CPU 6
#
#     see: /home/sfernandez/TOOLS/trinityrnaseq-Trinity-v2.5.0/sample_data/test_Trinity_Assembly/
#          for sample data and &#39;runMe.sh&#39; for example Trinity execution
#
#     For more details, visit: http://trinityrnaseq.github.io
#
###############################################################################</code></pre>
<p>Our work must be finished, let’s review the transcripts generated, which are in the file <code>Trinity.fasta</code>:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">head</span> trinity_out_dir/Trinity.fasta</code></pre></div>
<p>We note that the results are sequences in <code>fasta</code> format.</p>
<pre class="output"><code>&gt;TRINITY_DN8_c0_g1_i1 len=1720 path=[1:0-1719] [-1, 1, -2]
TTGCAATGCAAGTATTTAAGCGTATCACAACACATTGTTCTTCTCCAAGGTTCGTGAATC
GCTGTATTATTCTTTATTTTTCTTCAAGTGAGGATAAGAGTGATTGTTTGGCAAAGAAAA
ATTATGTTAACAAGTGTCTTATGGCCAAGGCACTTAAAGATTACCCCGTTCATACCAATA
TTGATCCTGATGCAGGGAAGTTATCATTTGACGATGCTTTTTACGAAGCTCACATTGAAC
TTCATTATCAATTTTTGAAGGAGGCTTCCCTAAATACCCTTATTAAAGATAAAAAAATGC
TCAAGTTTATTATCACTGTTCGTCCCGTTCATTTGCATGTCTCACCTTGGGTGGTTTATC
GTCGATATCGGGGGTTCAAAACTTTATACTATTTGTTAAAAAAGCAAAGTGCTAGAAATG
GGCGAGCTGTACCGAGTTTTCCTGTTTGGCGTGGAAACACGTATGAGAAGTTTCGTGAAG
GATTGTATTTTTTTATAGAAGCTTTACTGCATGATAGTCACTTTGCAACTAATGTTGATG</code></pre>
<h2 id="analyzing-the-statistics-of-the-assembled-transcriptome">Analyzing the statistics of the assembled transcriptome</h2>
<p>We can capture some statistics about this assembly using a program that is part of Trinity:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">$ <span class="kw">/usr/local/bin/trinityrnaseq/util/TrinityStats.pl</span> trinity_out_dirTrinityStats.pl trinity_out_dir/Trinity.fasta </code></pre></div>
<p>Which will generate the following data:</p>
<pre class="output"><code>################################
## Counts of transcripts, etc.
################################
Total trinity &#39;genes&#39;:  366
Total trinity transcripts:  374
Percent GC: 39.19

########################################
Stats based on ALL transcript contigs:
########################################

    Contig N10: 2605
    Contig N20: 2016
    Contig N30: 1596
    Contig N40: 1375
    Contig N50: 1090

    Median contig length: 419
    Average contig: 701.27
    Total assembled bases: 262274


#####################################################
## Stats based on ONLY LONGEST ISOFORM per &#39;GENE&#39;:
#####################################################

    Contig N10: 2576
    Contig N20: 1917
    Contig N30: 1591
    Contig N40: 1365
    Contig N50: 1075

    Median contig length: 405
    Average contig: 691.26
    Total assembled bases: 253002
</code></pre>
<p>This summary tells us:</p>
<ul>
<li>How many genes and transcripts were assembled</li>
<li>GC content</li>
<li>Statistics on the average size of the contigs</li>
<li>Number of bases assembled</li>
<li>Statistics on the average size of the longest isoform of each gene</li>
</ul>
<p>From this summary, the concept of Contig N50, N40, etc. is particularly important. For example, N50 indicates the size of the medium contig (or 50%) when all contigs are sorted by size. N40 is 40% etc. This measure can help to not simply rely on the longest contig and also allows us to observe how the overall length is increasing in all the transcripts.</p>
<p>Finally, we will perform a blast of the first 5 sequences to identify which organism they come from. Navigate to:</p>
<p><a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi">NCBI Blast</a></p>
<aside class="callout panel panel-info">
<div class="panel-heading">
<h2 id="whats-next"><span class="glyphicon glyphicon-pushpin"></span>What’s next?</h2>
</div>
<div class="panel-body">
<p>There are several things that can be done before taking this as the end result, such as:</p>
<ul>
<li>Check how complete the assembly is using <a href="https://busco.ezlab.org/">BUSCO</a>.</li>
<li>Annotate the genes.</li>
</ul>
</div>
</aside>
        </div>
      </div>
      </article>
      <div class="footer">
        <a class="label swc-blue-bg" href="mailto:selene.fernandez.valverde@gmail.com">Contact</a>
      </div>
    </div>
    <!-- Javascript placed at the end of the document so the pages load faster -->
    <script src="http://software-carpentry.org/v5/js/jquery-1.9.1.min.js"></script>
    <script src="css/bootstrap/bootstrap-js/bootstrap.js"></script>
    <script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>
  </body>
</html>
